storage-node/src/cache/lru_k.rs

```diff
@@ -0,0 +1,266 @@
+/// LRU-K cache implementation.
+/// Credit: https://doi.org/10.1145/170036.170081
+use core::panic;
+use datafusion::execution::cache;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+
+use super::CacheKey;
+use super::CacheValue;
+use super::{ParpulseCache, ParpulseCacheKey, ParpulseCacheValue};
+
+type Timestamp = i32;
+
+/// Represents a node in the LRU-K cache.
+///
+/// Each node contains a value of type `V` and a history of timestamps.
+/// The history is stored as a `VecDeque<Timestamp>`, where the most recent
+/// timestamps are at the front of the deque.
+struct LruKNode<V: CacheValue> {
+    value: V,
+    history: VecDeque<Timestamp>,
+}
+
+/// Represents an LRU-K cache.
+///
+/// The LRU-K algorithm evicts a node whose backward k-distance is maximum of all
+/// nodes. Backward k-distance is computed as the difference in time between current
+/// timestamp and the timestamp of kth previous access. A node with fewer than k
+/// historical accesses is given +inf as its backward k-distance. When multiple nodes
+/// have +inf backward k-distance, the cache evicts the node with the earliest
+/// overall timestamp (i.e., the frame whose least-recent recorded access is the
+/// overall least recent access, overall, out of all nodes).
+pub struct LruKCache<K: CacheKey, V: CacheValue> {
+    cache_map: HashMap<K, LruKNode<V>>,
+    max_capacity: usize,
+    size: usize,
+    curr_timestamp: Timestamp,
+    k: usize, // The k value for LRU-K
+}
+
+impl<K: CacheKey, V: CacheValue> LruKCache<K, V> {
+    pub fn new(max_capacity: usize, k: usize) -> LruKCache<K, V> {
+        LruKCache {
+            cache_map: HashMap::new(),
+            max_capacity,
+            size: 0,
+            curr_timestamp: 0,
+            k,
+        }
+    }
+
+    fn evict(&mut self, new_key: &K) -> Option<K> {
+        let mut found = false;
+        let mut max_k_dist = 0;
+        let mut k_dist = 0;
+        let mut earliest_timestamp = 0;
+        let mut key_to_evict: Option<K> = None;
+        for (key, node) in self.cache_map.iter() {
+            if key == new_key {
+                continue;
+            }
+            let history = &node.history;
+            if let Some(kth_timestamp) = history.front() {
+                k_dist = if history.len() < self.k {
+                    std::i32::MAX
+                } else {
+                    self.curr_timestamp - kth_timestamp
+                };
+                if (k_dist > max_k_dist)
+                    || (k_dist == max_k_dist && *kth_timestamp < earliest_timestamp)
```

Comment:
```suggestion
                if (k_dist > max_k_dist)
                    || (k_dist == max_k_dist && kth_timestamp < &earliest_timestamp)
```
A very small point: we can save one copy if we use `&earliest_timestamp` for comparison.